src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,150 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+    private Node<T> currentNode;
+    private Node<T> node;
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            head = new Node<>(null, value, tail);
+        } else if (size == 1) {
+            tail = new Node<>(head, value, null);
+            tail.prev = head;
+            head.next = tail;
+        } else {
+            currentNode = new Node<>(tail, value, null);
+            tail.next = currentNode;
+            tail = currentNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("This index is not valid!");
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        currentNode = getNode(index);
+        node = new Node<>(currentNode.prev, value, currentNode);
+        currentNode.prev = node;
+        if (node.prev != null) {
+            node.prev.next = node;
+        } else {
+            head = node;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        currentNode = getNode(index);
+        return currentNode.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        currentNode = getNode(index);
+        T returnValue = currentNode.item;
+        currentNode.item = value;
+        return returnValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        currentNode = getNode(index);
+        if (size == 1) {
+            size = 0;
+            return currentNode.item;
+        }
+        if (currentNode == head) {
+            head = currentNode.next;
+            head.prev = null;
+        } else if (currentNode == tail) {
+            tail = currentNode.prev;
+            tail.next = null;
+        } else {
+            currentNode.prev.next = currentNode.next;
+            currentNode.next.prev = currentNode.prev;
+        }
+        size--;
```

Comment:
let's move this part of code to the separate private method unlink() and then we can use this method in the remove by value method